smaug/operators/common.h

```diff
@@ -176,58 +271,109 @@ typedef unsigned ufx_t;
-// 8 packed 16-bit floating point values.
+/**  8 packed 16-bit floating point values. */
 typedef fp16_t v8ph_t
         __attribute__((__vector_size__(VECTOR_SIZE * sizeof(fp16_t))));
-// 4 packed 16-bit floating point values.
+/** 4 packed 16-bit floating point values. */
 typedef fp16_t v4ph_t
         __attribute__((__vector_size__(VECTOR_SIZE / 2 * sizeof(fp16_t))));
 
-// 8 packed 32-bit integer values.
+/** 8 packed 32-bit integer values. */
 typedef sfx_t v8sfx_t
         __attribute__((__vector_size__(VECTOR_SIZE * sizeof(sfx_t))));
+/** 4 packed 32-bit integer values. */
 typedef sfx_t v4sfx_t
         __attribute__((__vector_size__(VECTOR_SIZE / 2 * sizeof(sfx_t))));
 
-// 8 packed 8-bit bool values.
+/** 8 packed 8-bit bool values. */
 typedef uint8_t v8bl_t
         __attribute__((__vector_size__(VECTOR_SIZE * sizeof(uint8_t))));
 
-// Use these convenience macros to cast a raw pointer into a multidimensional
-// variable-length array, which lets us use [] notation inside of the ugly
-// sub2ind syntax!
-//
-// Usage:
-//   If we have an array like array[5][4]:
-//      ARRAY_2D(TYPE, output_name, array, 4);
-//
-//   If we have an array like array[5][4][3]:
-//      ARRAY_3D(TYPE, output_name, array, 4, 3);
-//
-//   If we have an array like array[5][4][3][2]
-//      ARRAY_4D(TYPE, output_name, array, 4, 3, 2);
-//
-//   And so on...
+/**
+ * Apply a elementwise mask to a 128-bit packed single precision FP vector.
+ *
+ * The mask is a vector of either 0s or -1s (all 1s). Entries that are have a
+ * mask of 0 are zeroed out.
+ *
+ * LLVM is smart enough to turn this into a SELECT instruction, rather than a
+ * bitwise mask!
+ *
+ * @param input A v4fp_t vector.
+ * @param mask A v4sfx_t vector of either 0s or -1s.
+ */
+#define VEC128_MASK(input, mask) ((v4fp_t)((v4sfx_t)input & mask))
+
+/**
+ * Same as VEC128_MASK, but for 256-bit vectors.
+ *
+ * @param input A v4fp_t vector.
+ * @param mask A v4sfx_t vector of either 0s or -1s.
+ */
+#define VEC256_MASK(input, mask) ((v8fp_t)((v8sfx_t)input & mask))
+
+/**
+ * @}
+ */
+
+/** \defgroup MultiDimArrays
+ *
+ * Use these convenience macros to cast a raw pointer into a multidimensional
+ * variable-length array, which lets us use `[]` notation instead of manually
+ * linearizing the index.
+ *
+ * Usage: Suppose we have an `int* array` pointer.
+ *
+ * To convert this into a multidimensional array, call the appropriate macro,
+ * providing the type, a new variable name (often the same as the original with
+ * an underscore prefixed), and then listing out all but the first dimension.
+ *
+ * For example:
+ *
+ *    ```c
+ *    // To convert into a 5x4 array:
+ *    ARRAY_2D(int, _array, array, 4);
+ *    _array[0][1] = 1;
+ *
+ *    // To convert into a 5x4x3 array:
+ *    ARRAY_3D(int, _array, array, 4, 3);
+ *    _array[0][1][2] = 1;
+ *
+ *    // To convert into a 5x4x3x2 array:
+ *    ARRAY_4D(int, _array, array, 4, 3, 2);
+ *    _array[0][1][2][3] = 1;
+ *    ```
+ * @{
```

Comment:
What's this for?

Reply:
It defines a Doxygen group, so that a group of related functions can be associated in the docs without having to repeat the same words.